### PR TITLE
Transformation: adding support to stretch a set of wide data frames into long data frames.

### LINKED ISSUE #35599: Grafana8: SQL data sources and worldmap panel issues
**What happened**:

My Grafana installation updated to version 8.0 and my visualizations broke.

**What you expected to happen**:

I expected the visualizations to continue working

**How to reproduce it (as minimally and precisely as possible)**:

I have asked about the problem on the [support forum](https://community.grafana.com/t/grafana-8-0-incompatibilities/48840). Got no useful replies; just one person said they had the same problem.

The forum post contains the actual query I am using. Note that it uses the `$__time(column)` macro. This is supposed to translate to `UNIX_TIMESTAMP(column) as time_sec`. In other words, the corresponding column ought to contain Unix epoch times. And, indeed, this is what MySQL returns, when the query is issued manually to it.

However, if I change the Worldmap visualization to a Table visualization, what I get in this column is human-readable timestamps. This probably confuses the Worldmap plug-in and it refuses to work.

The same problem occurs with the Pie Chart visualization - but I managed to fix it there by switching from time series to table and omitting the timestamp column entirely. Unfortunately, the Worldmap visualization requires time series, so I cannot use the same trick there.

Please, folks, this needs to be fixed ASAP. Currently all our visualizations are down, because they all contain world maps.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.0.1
- Data source type & version: MySQL
- OS Grafana is installed on: Ubuntu Linux 16.04
- User OS & Browser: Windows 7, Opera
- Grafana plugins: Worldmap, Pie Chart
- Others:

### PR DESCRIPTION
**Pull request by mckn**
**What this PR does / why we need it**:
This PR adds a transformation that will take a new "wide" data frame and stretch it into a "long" data frame. This really helpful to keep backward compatibility for panels that doesn't support the new wide data frame format.

We are planning on adding this to 8.0.6 and 7.5.10 to be able to support using the combination of new datasource plugins with panels not supporting the wide data frame format.

**Which issue(s) this PR fixes**:
Fixes #35599

**Special notes for your reviewer**:
I know that this isn't the optimal solution but I guess it is tricky to try to resolve the mapping issue located here: https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/dataframe/processDataFrame.ts#L326

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
can you move everything to grafana core, example: https://github.com/grafana/grafana/blob/main/public/app/core/components/TransformersUI/configFromQuery/configFromQuery.ts 

have the transform implementation split into two places makes things so messy :)

**Comment by torkelo** [comment]
They are never used from grafana/data or grafana/ui , they are always accessed from the registry, and the registry items are defined in core under TransformersUI so everything can be there

**Comment by torkelo** [comment]
Really tricky on what to name this transform :)

**Comment by mckn** [comment]
> Really tricky on what to name this transform :)

hehe, you didn't like the stretch name? :)

**Comment by mckn** [comment]
> They are never used from grafana/data or grafana/ui , they are always accessed from the registry, and the registry items are defined in core under TransformersUI so everything can be there

Love this! Great initiative 🙌🏻

**Review by ryantxu — COMMENTED** [review]
functionality wise... this looks great!

naming... needs work :(  but I'm not exactly sure what the names should be.  "long" frame means something else -- it is where duplicate times can exist.  We currently do not have a name for "array of frames with two fields: time,value"  cc:  @kylebrandt @sdboyer 

Related, we need a transformer to do type conversion from string > number columns.  I wonder if this all exists under the heading "Convert frame/field types"?

**Comment by leeoniya** [comment]
> Really tricky on what to name this transform :)

Unjoin?

**Comment by mckn** [comment]
> functionality wise... this looks great!
> 
> naming... needs work :( but I'm not exactly sure what the names should be. "long" frame means something else -- it is where duplicate times can exist. We currently do not have a name for "array of frames with two fields: time,value" cc: @kylebrandt @sdboyer
> 
> Related, we need a transformer to do type conversion from string > number columns. I wonder if this all exists under the heading "Convert frame/field types"?

I'm happy to change it to something better :)

**Comment by mckn** [comment]
> > Really tricky on what to name this transform :)
> 
> Unjoin?

Or ungroup?

**Comment by leeoniya** [comment]
maybe. unjoin is a weird word (is it a word?), but it is literally what this does, i think?

also it assumes all the time fields match each other when mixed?

**Comment by ryantxu** [comment]
We currently use the name "timeseries" to be anything with a time column.  In reality we have three types:

1. Wide timeseries
First time field, then any number of fields
Dimensions are attached as labels on fields
In uPlot, we also call this "aligned" data

2. Long timeseries (not supported by frontend)
duplicate timestamps are ok, and string fields are used to represent dimensions

3. [][time,number]  (no name yet)
This was the only thing supported in grafana 5
I have always though of it as just a wide timeseries with only one value -- but we need a better way to discuss this

**Comment by kylebrandt** [comment]
> We currently use the name "timeseries" to be anything with a time column. In reality we have three types:

There are examples of these types in the Go sdk docs under Frame https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go@v0.106.0/data#Frame , one example for each of the 3 shown.

**Comment by ryantxu** [comment]
Maybe a transformer called:  "Prepare time series", with options like:

1. Align data
This will do the same as join by time column.  I

2. Spread fields
(the behavior defined in this PR)
We can add help text that this was the only supported format before 6x

3. (not today) Wide to long
4. (not today) Long to wide

File: public/app/core/components/TransformersUI/stretchFrames/stretchFrames.test.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      }),
+    ];
+
+    expect(stretchFrames(source)).toEqual([
+      toEquableDataFrame({
+        name: 'wide0',
+        refId: 'A0',
+        fields: [
+          { name: 'time', type: FieldType.time, values: [10, 9, 8, 7, 6, 5] },
+          { name: 'text', type: FieldType.string, values: ['a', 'z', 'b', 'x', 'c', 'b'] },
```
Comment (ryantxu) [inline public/app/core/components/TransformersUI/stretchFrames/stretchFrames.test.ts:24]: I don't think we want to keep the string field -- old panels only expect numbers (that was all we could make)

**Comment by ryantxu** [comment]
One advantage of "prepare time series" is that we can also assert other constraints like sort order. and eventually mark it with metadata

**Comment by mckn** [comment]
Closing this in favor of #36737

**Pull request closed without merging**